### PR TITLE
#159204699 User rating articles

### PR DESCRIPTION
**Pull request by tutugodfrey**
#### What does this PR do?
- add the implementation for users to rate articles

#### Description of Task to be completed?
- write tests for the requirements for rating articles
- create a model (table) for storing rating data
- add routes to rating the articles by Id
- a middlewares function for abstraction
- add the controller to save ratings to the database
- add a route for getting article rating by id
- add a test for the get route

#### How should this be manually tested?
- clone the repository from https://github.com/andela/rivendell-ah
- run `npm install` and set up the environment variable based on the example in sample-env file
- run `npm test` to run test you can also interact with the application at
- http://localhost:3000/articles/rating/{articleId}?rating={ratingValue}

#### Any background context you want to provide?
- rating an article require Authorization token be provided in header

```
For reviewers:
files to review are  as listed below
/server/tests/article.tests.js
/server/controllers/articleControllers.js
review article rateArticle controller and getArticleRating in the articles controller
other modifitation in the article controller will be made by Jane when she is done with her feature.

```
#### What are the relevant pivotal tracker stories?
#159204699

#### Screenshots (if appropriate)
image
image
image
image
image
image

#### Questions:
N/A

### CODE REVIEW

File: server/controllers/RatingController.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                // add property to ratingInfo object
+                ratingInfo.ratingDetail.articleId = article.id;
+                ratingInfo.ratingDetail.authorId = article.authorId;
+                ratingInfo.rating = rateData;
+                return res.status(201).json(ratingInfo);
+              }
+              return res.status(403).json({
+                status: 'fail',
+                message: 'Your rating have already been recoded',
+              });
```
Comment (chidioguejiofor) [inline server/controllers/RatingController.js]: This is not our agreed convention for error response. The readme states:

```

{
    errors:{
         message: 'message',
    }
}
```

I think you should remove the `status: fail`

File: server/controllers/RatingController.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        include: [{ model: Rating, as: 'ratings' }],
+      })
+      .then((article) => {
+        if (article) {
+          if (article.authorId === userId) {
+            return res.status(403).json({
+              status: 'fail',
+              message: 'You can\'t provide ratings for your article',
+            });
+          }
```
Comment (chidioguejiofor) [inline server/controllers/RatingController.js]: I have noticed that this logic is repeated twice. I think you should abstract part of it to a new helper function so that you would increase `maintanability`

File: server/controllers/ArticleController.js [hunk 3]
```diff
@@ -135,11 +138,11 @@ class ArticleController {
       title, description, body,
     } = req.body.article;
     const authorId = req.body.decoded.id;
-    const updateField = [];
+    const fieldsArr = [];
 
-    if (title && title.trim()) updateField.push('title');
-    if (description && description.trim()) updateField.push('description');
-    if (body && body.trim()) updateField.push('body');
+    if (title) fieldsArr.push('title');
```
Comment (JCanaks) [inline server/controllers/ArticleController.js]: Great work but just a few things to note. The 'updateField' specified as the variable name is preferable over the 'fieldsArr' as specified in previous comments on the CRUD PR. The current changes should have been accepted for this file upon rebasing on the create article branch to avoid reverting back to old changes

File: server/utils/services/ratingServices.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+
+const ratingServices = {
+  /**
+   * @param {array} ratingArray array of ratings
+   * @returns {Object} object contain rating detailRating
+   */
+  computeRating(ratingArray) {
+    let sumOfRating = 0;
+    const ratingCount = ratingArray.length;
+    ratingArray.forEach((rating) => {
```
Comment (chidioguejiofor) [inline server/utils/services/ratingServices.js]: I thinkyou should remove this loop and use the SQL `AVE` function instead to calculate the average rating of the article.

File: server/controllers/ArticleController.js (hunk 3 above)
Comment (JCanaks) [reply server/controllers/ArticleController.js]: use git reflog to go back to how your branch was before rebase and then rebase again to accept the right changes

File: server/utils/services/ratingServices.js (hunk 4 above)
Comment (chidioguejiofor) [reply server/utils/services/ratingServices.js]: You can also use the `COUNT` function to get the number of ratings

File: server/controllers/RatingController.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+          // add property to ratingInfo object
+          ratingInfo.ratingDetail.articleId = article.id;
+          ratingInfo.ratingDetail.authorId = article.authorId;
+          ratingInfo.rating = article.ratings;
+          return res.status(200).json(ratingInfo);
+        }
+        return res.status(404).json({
+          status: 'fail',
+          message: 'Article not found',
```
Comment (chidioguejiofor) [inline server/controllers/RatingController.js]: Same convention issue here as above

File: server/database/models/rating.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+module.exports = (sequelize, DataTypes) => {
+  const Rating = sequelize.define('Rating', {
+    rating: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+    },
+    userId: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+    },
```
Comment (chidioguejiofor) [inline server/database/models/rating.js]: Is `userId` not suppose to reference `Users` table here or is there a reason why you decided to do this

File: server/database/models/rating.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      },
+    },
+  }, {});
+
+  Rating.associate = (models) => {
+    // associate rating with article
+    Rating.belongsTo(models.Article, {
+      foreignKey: 'articleId',
+      onDelete: 'CASCADE',
+    });
```
Comment (chidioguejiofor) [inline server/database/models/rating.js]: Similar to the comment I made above, Is rating not supposed to belong to a user or do you have any reason for not doing this?

**Review by chidioguejiofor — CHANGES_REQUESTED** [review]
Nice work completing your feature but I found some places where you did not follow conventions. Also, I think you need to checkout you relationships between `Raing` and `User`

File: server/controllers/RatingController.js [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+                ratingInfo.ratingDetail.authorId = article.authorId;
+                ratingInfo.rating = rateData;
+                return res.status(201).json(ratingInfo);
+              }
+              return res.status(403).json({
+                status: 'fail',
+                message: 'Your rating have already been recoded',
+              });
+            });
+        }
```
Comment (JCanaks) [inline server/controllers/RatingController.js]: as mentioned in previous PR comments the property status with value 'fail' is to be removed and any uncessesary return messages for success as well

File: server/controllers/RatingController.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        where: {
+          slug,
+        },
+        include: [{ model: Rating, as: 'ratings' }],
+      })
+      .then((article) => {
+        if (article) {
+          if (article.authorId === userId) {
+            return res.status(403).json({
+              status: 'fail',
```
Comment (JCanaks) [inline server/controllers/RatingController.js]: The status property here as well should be removed and in all instances where it has been used

File: server/database/migrations/20180807123711-create-rating.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      rating: {
+        allowNull: false,
+        type: Sequelize.INTEGER,
+      },
+      userId: {
```
Comment (chidioguejiofor) [inline server/database/migrations/20180807123711-create-rating.js]: I think `userId` should reference `Users` table

File: server/database/models/rating.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      type: DataTypes.UUID,
+      defaultValue: DataTypes.UUIDV4,
+      allowNull: false,
+      onDelete: 'CASCADE',
+      reference: {
+        model: 'Article',
+        key: 'id',
+      },
+    },
+  }, {});
```
Comment (chidioguejiofor) [inline server/database/models/rating.js]: I think you should index the `articleId` and `userId` to be `unique` to ensure that the combination of userid and articleid are unique and cannot be duplicated

For instance, 
```
 {
    indexes: [
      {
        unique: true,
        fields: ['articleId', 'userId'],
      },
    ],
  }
```

Comment (chidioguejiofor) [reply server/database/models/rating.js]: The above code should be in the second argument of the model declaration

**Review by chidioguejiofor — CHANGES_REQUESTED** [review]

File: server/database/models/rating.js (hunk 6 above)
Comment (JCanaks) [reply server/database/models/rating.js]: There should be an association between the user and the ratings table

File: server/utils/middleware/ratingMiddleware.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+const ratingMiddleware = {
+  /**
+   * @param {*} req the request body
+   * @param {*} res the res body
+   * @param {*} next a call to the next middle function
+   */
+  validateRating(req, res, next) {
+    const numberFormat = /^\d*[1-9]\d*$/;
+    if (
+      (numberFormat.test(req.params.slug))
```
Comment (megame24) [inline server/utils/middleware/ratingMiddleware.js]: Why are you testing the slug?

File: server/database/models/rating.js (hunk 7 above)
Comment (tutugodfrey) [reply server/database/models/rating.js]: Ok, thanks for the observation. I think the association that is necessary here is the relationship between Rating and Article. Since there is no need to return the users along with the ratings but rather the article

File: server/utils/middleware/ratingMiddleware.js (hunk 12 above)
Comment (tutugodfrey) [reply server/utils/middleware/ratingMiddleware.js]: I am testing to prevent digit being entered as the slug. i.e 1, 4 etc.

**Review by JCanaks — APPROVED** [review]

**Review by chidioguejiofor — APPROVED** [review]

**Review by mppdg — APPROVED** [review]

**Review by megame24 — APPROVED** [review]

File: server/test/routes/oauth.test.js [hunk 13]
```diff
@@ -41,3 +42,6 @@ describe('Social Login', () => {
     }).timeout(50000);
   });
 });
+};
+
+export default oAuthTest;
```
Comment (rhenshaw56) [inline server/test/routes/oauth.test.js]: why are you exporting the test ?

File: server/test/routes/index.test.js [hunk 14]
```diff
@@ -17,4 +28,11 @@ describe('Testing the index.js file', () => {
         });
     });
   });
+  authTest();
```
Comment (rhenshaw56) [inline server/test/routes/index.test.js]: why are you exporting all the tests ?

File: server/test/routes/auth.test.js [hunk 15]
```diff
@@ -22,6 +22,7 @@ const tempUser = {
     password: '11qQw@123',
   },
 };
+const authTest = () => {
```
Comment (rhenshaw56) [inline server/test/routes/auth.test.js]: you don't need this assignment, please remove it

File: server/controllers/RatingController.js [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+                obj.averageRating = obj.averageRating.toPrecision(3);
+                return res.status(200).json({
+                  ratingDetails: obj,
+                  rating: updatedRating[0],
+                });
+              });
+            }
+            return res.status(404).json({
+              errors: {
+                message: 'Raiting not found, no update was made',
```
Comment (rhenshaw56) [inline server/controllers/RatingController.js]: `Rating` not `Raiting`

File: server/controllers/RatingController.js [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+                attributes: [
+                  [sequelize.fn('SUM', sequelize.col('rating')),
+                    'totalRating'],
+                  [sequelize.fn('COUNT', sequelize.col('rating')),
+                    'ratingCount'],
+                  [sequelize.fn('AVG', sequelize.col('rating')),
+                    'averageRating'],
+                ],
+              }).then((ratingStats) => {
+                const obj = ratingStats.dataValues;
```
Comment (rhenshaw56) [inline server/controllers/RatingController.js]: what is `obj` ? please rename the variable to be representative of what it is

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

File: server/test/routes/oauth.test.js (hunk 13 above)
Comment (tutugodfrey) [reply server/test/routes/oauth.test.js]: I am exporting the test files because of this configuration
`  after((done) => {
    models.sequelize.close();
  done();
  });
`
in test/routes/users.test that was blocking some test files from running. I move the configuration to test/index.test. So exporting the test files and importing them to index.test file will allow us to specify the order the tests should run

File: server/test/routes/auth.test.js (hunk 15 above)
Comment (tutugodfrey) [reply server/test/routes/auth.test.js]: I need to be more clear, is it the 
`const authTest = ()`
or the entire function assignment
`const authTest = () => {`

File: package.json [hunk 18]
```diff
@@ -12,8 +12,8 @@
     "dev": "NODE_ENV=development nodemon ./server/index.js --exec babel-node",
     "lint": "eslint .",
     "pretest": "npm run posttest && sequelize --env test db:migrate && sequelize --env test db:seed:all",
-    "test": "cross-env NODE_ENV=test nyc mocha --timeout=8000 --compilers js:babel-core/register ./server/test/**/*.js",
-    "coverage": "npm run pretest && cross-env NODE_ENV=test babel-node ./node_modules/babel-istanbul/lib/cli.js cover ./node_modules/mocha/bin/_mocha --timeout=8000 ./server/test/**/*.js && npm run posttest ",
+    "test": "cross-env NODE_ENV=test nyc mocha --timeout=28000 --compilers js:babel-core/register ./server/test/**/*.js",
```
Comment (rhenshaw56) [inline package.json]: change back to `8000`

File: server/test/routes/index.test.js [hunk 19]
```diff
@@ -4,9 +4,9 @@ import chai, { expect, use, request } from 'chai';
 import chaiHttp from 'chai-http';
 import server from '../../index';
 
-use(chaiHttp);
+chai.use(chaiHttp);
 
-describe('Testing the index.js file', () => {
+describe('Tests', () => {
```
Comment (rhenshaw56) [inline server/test/routes/index.test.js]: what happened to `Testing the index.js file` ?

File: server/test/routes/likes.test.js [hunk 20]
```diff
@@ -14,6 +14,7 @@ let slug;
 let likeRoute;
 let unverifiedToken;
 
+// const articleLikeTest = () => {
```
Comment (rhenshaw56) [inline server/test/routes/likes.test.js]: get rid of the commented code

File: server/test/routes/oauth.test.js [hunk 21]
```diff
@@ -6,6 +6,7 @@ import chaiHttp from 'chai-http';
 import server from '../../index';
 
 use(chaiHttp);
+// const oAuthTest = () => {
```
Comment (rhenshaw56) [inline server/test/routes/oauth.test.js]: get rid of the commented bit

File: server/test/routes/users.test.js [hunk 22]
```diff
@@ -56,8 +56,8 @@ describe('Testing user routes', () => {
         .end((err, res) => {
           expect(res.status).to.equal(201);
           expect(res.body).to.haveOwnProperty('message')
-            .to.equal(`Sign up successful, visit your email 
-          to verify your account.`);
+            .to.equal(
+              `Sign up successful, visit your email to verify your account.`);
```
Comment (rhenshaw56) [inline server/test/routes/users.test.js]: did you write this test ? if so why the change ? pls only touch what's related to your current task

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

**Review by rhenshaw56 — APPROVED** [review]
